File: lib/go/thrift/outbound_stream.go
package thrift

import "io"

type TOutboundStream interface {
	io.Closer

	Send(Context, TRequest) error
}

type tOutboundStream struct {
	tBaseStream

	messageType TMessageType
}

func newTClientOutboundStream(name string, seqID int32, in, out TProtocol, cl *TSyncClient) *tOutboundStream {
	return &tOutboundStream{
		tBaseStream: newTClientBaseStream(name, seqID, in, out, CLIENT_STREAM_GOAWAY, cl),
		messageType: CLIENT_STREAM_MESSAGE,
	}
}

func newTServerOutboundStream(name string, seqID int32, in, out TProtocol) *tOutboundStream {
	return &tOutboundStream{
		tBaseStream: newTServerBaseStream(name, seqID, in, out, SERVER_STREAM_GOAWAY),
		messageType: SERVER_STREAM_MESSAGE,
	}
}

func (s *tOutboundStream) readGoaway() {
	mt, err := s.readShell()

	if err != nil {
		s.close()
		return
	}

	if mt == s.goAwayType {
		s.writeGoAwayACK()
	}

	s.close()
}

func (s *tOutboundStream) ready() {
	s.readyOnce.Do(func() {
		close(s.readyc)
		go s.readGoaway()
	})
}

func (s *tOutboundStream) Close() error {
	select {
	case <-s.readyc:
	case <-s.closec:
		return nil
	}

	select {
	case <-s.closec:
		return nil
	default:
	}

	if err := s.writeGoAway(); err != nil {
		s.close()
		return err
	}

	<-s.closec
	return nil
}

func (s *tOutboundStream) Send(ctx Context, req TRequest) error {
	return s.write(ctx, s.messageType, req)
}

File: lib/rb/lib/thrift/outbound_stream.rb
module Thrift
  class TOutboundStream < TBaseStream
    def initialize(iprot, oprot, name, seqid, klass, message_type, unlock = nil)
      @message_type = message_type
      @klass = klass
      @unlock = unlock
      super(iprot, oprot, name, seqid)
    end

    def ready
      Thread.new { read_goaway }
      super
    end

    def close
      wait_ready

      write_shell @message_type + 1

      @mutex.synchronize do
        loop do
          @cond.wait(@mutex)

          return nil if @closed
        end
      end
    end

    def send(arg)
      raise EOFError if @closed

      unless @oprot.trans.open?
        mark_as_closed
        raise EOFError
      end

      wait_ready
      write @message_type, @klass.new(arg: arg)
    end

    private

    def read_goaway
      mtype = read_shell

      write_shell @message_type + 2 if mtype == (@message_type + 1)
    ensure
      mark_as_closed
      @unlock&.call
    end
  end
end
